tests/unit/s2n_self_talk_psk_test.c

```diff
@@ -0,0 +1,340 @@
+    GUARD_AS_RESULT(s2n_psk_new_secret(other_server_psk, other_server_data, sizeof(other_server_data)));
+    other_server_psk->hmac_alg = S2N_HMAC_SHA384;
+
+    return S2N_RESULT_OK;
+}
+
+static int s2n_test_select_psk_identity_callback(struct s2n_connection *conn, struct s2n_psk_identity *identities,
+                                                 size_t identities_length, uint16_t *chosen_wire_index)
+{
+    notnull_check(conn);
+    notnull_check(identities);
+    notnull_check(chosen_wire_index);
+
+    struct s2n_array *known_psks = &conn->psk_params.psk_list;
+
+    for (size_t i = 0; i < identities_length; i++) {
+        struct s2n_blob wire_identity = { 0 };
+        GUARD(s2n_blob_init(&wire_identity, identities[i].data, identities[i].length));
+
+        struct s2n_psk *local_match = NULL;
+        GUARD_AS_POSIX(s2n_match_psk_identity(known_psks, &wire_identity, &local_match));
+
+        if (local_match != NULL) {
+            *chosen_wire_index = i;
+            return S2N_SUCCESS;
+        }
+    }
+
+    return S2N_FAILURE;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Happy Case: PSK without callback function */ 
+    {
+        /* Setup connections */
+        struct s2n_connection *client_conn = NULL;
+        struct s2n_connection *server_conn = NULL;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        struct s2n_config *config = NULL;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Setup PSKs */
+        EXPECT_OK(setup_client_psks(client_conn));
+        EXPECT_OK(setup_server_psks(server_conn));
+
+        /* Negotiate handshake */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Validate handshake type */
+        EXPECT_FALSE(client_conn->handshake.handshake_type & CLIENT_AUTH);
+        EXPECT_FALSE(server_conn->handshake.handshake_type & CLIENT_AUTH);
```

Comment:
The handshake wouldn't be CLIENT_AUTH even if you weren't using PSKs. Client auth has to be explicitly turned on. Which is actually probably a good test case; we expect client auth to be ignored when using PSKs.